### PR TITLE
Add support for lowering linalg_ext.scatter to loops + Linalg ops.

### PR DESCRIPTION
**Pull request by hanhanW**
This supports both scalar and slice cases. For slice cases, we will
create a linalg.generic op with parallel loops to handle slices. It also
works for cases that index_depth > 1.

This is a step towards https://github.com/google/iree/issues/6403

### CODE REVIEW

**Comment by hanhanW** [comment]
@MaheshRavishankar I think the scf.for body creation is related to TiedOpInterface. We probably want to move them to be part of `getTiledImplementation`.

To me, it looks like we create memref.load/subview the operands and pass it to an implementation. E.g., in slice cases, we will apply subview on `updates` based on iv and apply subview on `original` based on loaded indices. The former subview is part of how we tile the operation. The latter subview + linalg.generic op is how we implement tiling Linalg ops. We probably will borrow the logic to getTiledImplementation of `linalg_ext.scatter`.

**Comment by MaheshRavishankar** [comment]
> @MaheshRavishankar I think the scf.for body creation is related to TiedOpInterface. We probably want to move them to be part of `getTiledImplementation`.
> 
> To me, it looks like we create memref.load/subview the operands and pass it to an implementation. E.g., in slice cases, we will apply subview on `updates` based on iv and apply subview on `original` based on loaded indices. The former subview is part of how we tile the operation. The latter subview + linalg.generic op is how we implement tiling Linalg ops. We probably will borrow the logic to getTiledImplementation of `linalg_ext.scatter`.

I am not sure I follow the comment here. We definitely need these ops to have a `TiedOpInterface`, but `LinalgOp`s dont have this interface. Instead there is some special logic in `DispatchLinalgOnTensors` pass and `LinalgBufferize` pass that implements this behavior. The `LinalgExt` op should do the same. Ill take a pass at this sometime (currently implementing the `TiledOpInterface`, so don't have cycles for this)

File: iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp [hunk 1]
```diff
@@ -119,23 +124,166 @@ struct BubbleSortConversion : public OpRewritePattern<linalg_ext::SortOp> {
...
+
+struct ScatterScalarConversion
+    : public OpRewritePattern<linalg_ext::ScatterOp> {
+  using OpRewritePattern<linalg_ext::ScatterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg_ext::ScatterOp op,
+                                PatternRewriter& rewriter) const final {
+    if (!op.hasBufferSemantics()) return failure();
+
+    auto updates = op.getInputOperand(0);
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp:139]: Nit: I think the `scatter` op has `updates()`, `indices()` and `original()` methods to do this. We could use that

File: iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir [hunk 2]
```diff
@@ -94,3 +94,175 @@ func @sort_multi(%arg0: memref<128xf32>, %arg1: memref<128xi32>) {
 // CHECK:               memref.store %[[V4]], %[[BUF2]][%[[ARG2]]]
 // CHECK:               memref.store %[[V3]], %[[BUF2]][%[[T2]]]
 // CHECK:             }
+
+// -----
+
+func @scatter_update_scalar_1D(
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir:100]: Nit: Could we make all these tests work with dynamic shapes. Its a bit more of Lit testing goo, but it keeps us more honest in terms of dynamic shape codegen.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a minor nit, but overall looks fine.

File: iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp (hunk 1 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp:139]: Good point, let me add the methods to linalg_ext.scatter op!

File: iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir (hunk 2 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir:100]: Sure, let me add more tests for dynamic shapes.

File: iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp (hunk 1 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/Transforms/ConvertToLoops.cpp]: I also added more methods, like `getUpdateSliceRank`, `isScalarUpdate`.

File: iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir (hunk 2 above)
Comment (hanhanW) [reply iree/compiler/Dialect/LinalgExt/Transforms/test/convert_to_loops.mlir:100]: Added three more test cases, which covers `index_depth=1/scalar_update` case, `index_depth=2/scalar_update` case, and `index_depth=1/slice_update` case.

File: iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td [hunk 3]
```diff
@@ -83,6 +83,26 @@ def LinalgExt_ScatterOp : LinalgExt_Op<"scatter"> {
           .getShape()
           .back();
     }
+
+    Value updates() {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.td:87]: Thanks for adding these. After I mentioned it, I realized I hadnt actually added this. It was part of my local changes. This looks good.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**